lib/executor/constant_arrival_rate_test.go

```diff
@@ -0,0 +1,142 @@
+package executor
+
+import (
+	"context"
+	"io/ioutil"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/testutils"
+	"github.com/loadimpact/k6/lib/types"
+	"github.com/loadimpact/k6/stats"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	null "gopkg.in/guregu/null.v3"
+)
+
+func testConstantArrivalRateSetup(t *testing.T, vuFn func(context.Context, chan<- stats.SampleContainer) error) (
+	context.Context, context.CancelFunc, lib.Executor, *testutils.SimpleLogrusHook) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var config = ConstantArrivalRateConfig{
+		TimeUnit:        types.NullDurationFrom(time.Second),
+		Rate:            null.IntFrom(50),
+		Duration:        types.NullDurationFrom(5 * time.Second),
+		PreAllocatedVUs: null.IntFrom(10),
+		MaxVUs:          null.IntFrom(20),
+	}
+	logHook := &testutils.SimpleLogrusHook{HookedLevels: []logrus.Level{logrus.WarnLevel}}
+	testLog := logrus.New()
+	testLog.AddHook(logHook)
+	testLog.SetOutput(ioutil.Discard)
+	logEntry := logrus.NewEntry(testLog)
+	es := lib.NewExecutionState(lib.Options{}, 10, 50) //TODO: fix
+	runner := lib.MiniRunner{
+		Fn: vuFn,
+	}
+
+	es.SetInitVUFunc(func(_ context.Context, _ *logrus.Entry) (lib.VU, error) {
+		return &lib.MiniRunnerVU{R: runner}, nil
+	})
+
+	initializeVUs(ctx, t, logEntry, es, 10)
+
+	executor, err := config.NewExecutor(es, logEntry)
+	require.NoError(t, err)
+	err = executor.Init(ctx)
+	require.NoError(t, err)
+	return ctx, cancel, executor, logHook
+}
+
+func TestConstantArrivalRateRunNotEnoughAlloctedVUsWarn(t *testing.T) {
+	t.Parallel()
+	var ctx, cancel, executor, logHook = testConstantArrivalRateSetup(
+		t, func(ctx context.Context, out chan<- stats.SampleContainer) error {
+			time.Sleep(time.Second)
+			return nil
+		})
+	defer cancel()
+	var engineOut = make(chan stats.SampleContainer, 1000)
+	err := executor.Run(ctx, engineOut)
+	require.NoError(t, err)
+	entries := logHook.Drain()
+	require.NotEmpty(t, entries)
+	for _, entry := range entries {
+		require.Equal(t,
+			"Insufficient VUs, reached 20 active VUs and cannot allocate more",
+			entry.Message)
+		require.Equal(t, logrus.WarnLevel, entry.Level)
+	}
+}
+
+func TestConstantArrivalRateRunCorrectRate(t *testing.T) {
+	t.Parallel()
+	var count int64
+	var ctx, cancel, executor, logHook = testConstantArrivalRateSetup(
+		t, func(ctx context.Context, out chan<- stats.SampleContainer) error {
+			atomic.AddInt64(&count, 1)
+			return nil
+		})
+	defer cancel()
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// check that we got around the amount of VU iterations as we would expect
+		var currentCount int64
+
+		for i := 0; i < 5; i++ {
+			time.Sleep(time.Second)
+			currentCount = atomic.SwapInt64(&count, 0)
+			require.InDelta(t, 50, currentCount, 1)
```

Comment:
I'm a bit worried that this test might be flaky under load

Reply:
this one I am less worried then the other [one](https://github.com/loadimpact/k6/pull/1117/files#diff-1237bf320cc289c1f30cdb59a7753ecdR287-R322), but both should be ... a lot less flaky then the ones we have been fixing so far ... 
If you want I can drop the t.Parallel for those ?

Reply:
No, I vote for leaving it as it is for now, let's wait and see if they start failing.... If that happens, we can easily relax the constraints, but it would would give us some information how robust the scheduling code is under load.